### ISSUE TITLE
Modern wrapper plan

### ISSUE THREAD

**Issue by dellaert** [borglab/gtsam#204]
Desired end-state: python scripts generate both python and MATLAB wrappers. Because there is an API change in python, this will bump our version to 4.1.

- [x] PR 0 to package *existing* cython wrapper for manylinux, README, so loads from colab (this is done in a separate repo, https://github.com/borglab/gtsam-manylinux-build)
- [x] Add pybind to feature/new_wrapper
- [x] Add pybind wrapper version (purely additive, duplicates all python scripts/tests into gtsam_py, includes GTSAM changes), ~fix Values~ (done), PR to feature/new_wrapper
- [x] Add travis CI for cython wrapper, PR to develop
- [x] Add travis CI for new wrapper, PR to feature/new_wrapper
- [x] Do PR to develop with new wrapper and we ask everyone to test/switch
- [x] @ProfFan tests MATLAB wrapper 
- [x] Switch MATLAB wrapper to be created by python, includes Matthew unit tests, PR to feature/new_wrapper
- [x] Fix most Bugs before 4.1
- [x] Remove cython wrapper and old c++ wrap, PR to develop (only file deletions)
- [ ] Create release branch and do PR to master, add release tag
- [ ] Create packaging in gtsam-manylinux-build for 4.1 so we can pip install, think about versioning

**Comment by ProfFan** [comment]
@dellaert Now gtsam works in Colab with some patches :)

Demo: https://colab.research.google.com/drive/1j2HgxVUDUfQEseLmke4Tk_LKiyJ87QjQ

One blocker: pypi got a file size limit of 60MB, but GTSAM packaged is 272M... I filed a request for increase the quota, in the meantime can load from Google Drive or an URL instead.

**Comment by ProfFan** [comment]
Code to build the wheels:

https://github.com/ProfFan/gtsam-manylinux-build

**Comment by ProfFan** [comment]
@dellaert Now GTSAM lives on PyPI with a size of 7.3MB :)

I changed the Colab notebook to just use `pip install gtsam` :smile:

**Comment by dellaert** [comment]
This is *AMAZING*

**Comment by dellaert** [comment]
How did you get it down to 7.3MB??

**Comment by ProfFan** [comment]
I stripped out the debug symbols and used a custom boost build instead of the system provided package...

**Comment by ProfFan** [comment]
@dellaert Note that macOS build is complete and tested (Python 3.7).

**Comment by dellaert** [comment]
> @dellaert Note that macOS build is complete and tested (Python 3.7).

Does that mean we can now move on to "Add pybind wrapper version, also with CI support (purely additive, duplicates all python scripts/tests into gtsam_py, includes GTSAM changes)" ?

**Comment by ProfFan** [comment]
Yes.

**Comment by dellaert** [comment]
@ProfFan Did you add Travis CI for (both) wrappers in your branch yet? I think that's crucial before we can merge into develop.

**Comment by dellaert** [comment]
@ProfFan I checked out the branch but confused about flags. There is `GTSAM_INSTALL_CYTHON_TOOLBOX` and `GTSAM_BUILD_PYTHON`. Should the latter be `GTSAM_INSTALL_PYTHON_TOOLBOX` ?

**Comment by ProfFan** [comment]
I kept them this way because that is what duy used. After all, I think the first `INSTALL` should be renamed to `GTSAM_BUILD_CYTHON_WRAPPER` as this flags actually does not control the installation of the wrapper. It only builds the wrapper.

**Comment by ProfFan** [comment]
@dellaert Yes, but I haven't added the actual tests, just pure compile (and package to wheel). I will add it really quick.

**Comment by dellaert** [comment]
I agree, but I propose we are consistent with the (slightly wrong) past, and use `GTSAM_INSTALL_PYTHON_TOOLBOX`

**Comment by ProfFan** [comment]
@dellaert Got it. I'll go that way  :)

**Comment by chrisbeall** [comment]
What is the current state of the pybind wrapper? Is it ready for testing in `feature/new_wrapper`, or are there still significant missing pieces? We are very interested in trying this at Nuro :-)

**Comment by dellaert** [comment]
@ProfFan good idea: please give status update and update TODO list above.

**Comment by ProfFan** [comment]
@chrisbeall @dellaert It is in a *kind-of* final state-except it is blocked by the Travis timeouts because the generated `gtsam.cpp` is too large. It builds (and tests) nicely on GitHub Actions, and that is the reason behind #259 :)

**Comment by dellaert** [comment]
Can you update checkmarks above? What's missing ?

**Comment by dellaert** [comment]
@chrisbeall consider commenting/approving #259 :-)

**Comment by dellaert** [comment]
@ProfFan two remaining items after mega-merge:

- [ ] Can we now make typedef default and simply remove conditional code?
- [ ]  After that's done, please tag a release 4.1rc, and create a corresponding python wheel?

**Comment by BioGeek** [comment]
Small note, in the [Colab notebook](https://colab.research.google.com/drive/1j2HgxVUDUfQEseLmke4Tk_LKiyJ87QjQ) by @ProfFan  [mentioned earlier](https://github.com/borglab/gtsam/issues/204#issuecomment-571344287), you have to change `gtsam.noiseModel_Diagonal.Sigmas`  to `gtsam.noiseModel.Diagonal.Sigmas` to get it to run correctly.
